lib/uklibid/include/uk/libinfo.h

```diff
@@ -0,0 +1,87 @@
+
+#include <uk/arch/types.h>
+#include <uk/essentials.h>
+
+/*
+ * Library information is stored in the special section " .uk_libinfo". For
+ * each library there is one `struct uk_libid_info_hdr` generated. Each header
+ * can contain multiple `struct uk_libid_info_rec` records. A library name
+ * record (`UKLI_REC_LIBNAME`) is the only mandatory record type for each
+ * library.
+ * Additonally, a single header is created for global information, which differs
+ * from the library headers in that it does not contain a record for a library
+ * name.
+ *
+ * Layout for each information header:
+ *
+ *                              _____________
+ *  +--------------------------+             \
+ *  | struct uk_libid_info_hdr |              |
+ *  |                          |               > hdr->len
+ *  +--------------------------+              |
+ *  | struct uk_libid_info_rec |              |
+ *  +--------------------------+ -.           |
+ *  | struct uk_libid_info_rec |   > rec->len |
+ *  |                          |  |           |
+ *  +--------------------------+ -`           |
+ *  :          . . .           :              :
+ *  +--------------------------+              |
+ *  | struct uk_libid_info_rec |              |
+ *  |                          |              |
+ *  +--------------------------+ ____________/
+ */
+
+/*
+ * NOTE: The layout version needs only be updated if the structs
+ *       `uk_libid_info_rec` and `uk_libid_info_hdr` are updated.
+ */
+ #define UKLI_LAYOUT              0x0001
+
+/*
+ * NOTE: In order to keep backwards compatibility with pre-compiled
+ *       libraries, this list should only be extended with new entries.
+ */
+#define UKLI_REC_LIBNAME          0x0001 /* C-string, absent for global hdr */
+#define UKLI_REC_COMMENT          0x0002 /* C-string */
+#define UKLI_REC_VERSION          0x0003 /* C-string */
+#define UKLI_REC_LICENSE          0x0004 /* C-string */
+#define UKLI_REC_GITDESC          0x0005 /* C-string */
+#define UKLI_REC_UKVERSION        0x0006 /* C-string */
+#define UKLI_REC_UKFULLVERSION    0x0007 /* C-string */
+#define UKLI_REC_UKCODENAME       0x0008 /* C-string */
+#define UKLI_REC_UKCONFIG         0x0009 /* binary data */
+#define UKLI_REC_UKCONFIGGZ       0x000A /* binary data */
+#define UKLI_REC_COMPILER         0x000B /* C-string */
+#define UKLI_REC_COMPILEDATE      0x000C /* C-string */
+#define UKLI_REC_COMPILEDBY       0x000D /* C-string */
+#define UKLI_REC_COMPILEDBYASSOC  0x000E /* C-string */
+
+#if !__ASSEMBLY__
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct uk_libid_info_rec {
+	unsigned short type;    /* record type */
+	unsigned int   len;     /* including record header */
+	char           data[];  /* raw data */
+} __packed __align(1);
+
+/*
+ * NOTE: Due to binary cross-compatibility, the sizes, offsets, data types and
+ *       meanings must remain the same regardless of the layout version:
+ *       `len` and `version`.
+ */
```

Comment:
I thought about this, too, both actually. My problem (but it isn't an excuse) was that we do not have fix width integer definitions in assembler, we would need to provide them with `<uk/arch/types.h>`. Because of the additional requirement of endianess, I thought the way to go about it could be by detecting the image architecture (e.g., parsing ELF headers). Both information (integer width and endianess) can be derived from this for external tools and internally we have the types matching anyways. What do you think? I am okay with adding assembler equivalent types for `__u16`, `__u8`, in fact this may make dealing with binary image formats easier, when there is no simple header for detecting the architecture.

Reply:
I updated the code to introduce assembler equivalents for `__u8`, `_u16`, `__u32` and so on...